applications/asset_tracker/src/cloud_codec/cloud_codec.c

```diff
@@ -0,0 +1,385 @@
+
+static int json_add_obj(cJSON *parent, const char *str, cJSON *item)
+{
+	cJSON_AddItemToObject(parent, str, item);
+
+	return 0;
+}
+
+static int json_add_num(cJSON *parent, const char *str, double num)
+{
+	cJSON *json_num;
+
+	json_num = cJSON_CreateNumber(num);
+	if (json_num == NULL) {
+		return -ENOMEM;
+	}
+
+	return json_add_obj(parent, str, json_num);
+}
+
+static int json_add_str(cJSON *parent, const char *str, const char *item)
+{
+	cJSON *json_str;
+
+	json_str = cJSON_CreateString(item);
+	if (json_str == NULL) {
+		return -ENOMEM;
+	}
+
+	return json_add_obj(parent, str, json_str);
+}
+
+static int json_add_null(cJSON *parent, const char *str)
+{
+	cJSON *json_null;
+
+	json_null = cJSON_CreateNull();
+	if (json_null == NULL) {
+		return -ENOMEM;
+	}
+
+	return json_add_obj(parent, str, json_null);
+}
+
+static cJSON *json_object_decode(cJSON *obj, const char *str)
+{
+	return obj ? cJSON_GetObjectItem(obj, str) : NULL;
+}
+
+int cloud_encode_data(const struct cloud_channel_data *channel,
+			struct cloud_data *output)
+{
+	int ret;
+
+	__ASSERT_NO_MSG(channel != NULL);
+	__ASSERT_NO_MSG(channel->data.buf != NULL);
+	__ASSERT_NO_MSG(channel->data.len != 0);
+	__ASSERT_NO_MSG(output != NULL);
+
+	cJSON *root_obj = cJSON_CreateObject();
+
+	if (root_obj == NULL) {
+		return -ENOMEM;
+	}
+
+	ret = json_add_str(root_obj, "appId", channel_type_str[channel->type]);
+	ret += json_add_str(root_obj, "data", channel->data.buf);
+	ret += json_add_str(root_obj, "messageType", "DATA");
+
+	if (ret != 0) {
+		cJSON_Delete(root_obj);
+		return -ENOMEM;
+	}
+
+	char *buffer;
+
+	buffer = cJSON_PrintUnformatted(root_obj);
+	cJSON_Delete(root_obj);
+
+	output->buf = buffer;
+	output->len = strlen(buffer);
+
+	return 0;
+}
+
+static int cloud_search_cmd(cJSON *group_obj, enum cloud_cmd_group group)
+{
+	__ASSERT_NO_MSG(group_obj != 0);
+
+	cJSON *recipient_obj	= NULL;
+	cJSON *channel_obj	= NULL;
+	cJSON *type_obj		= NULL;
+
+	struct cmd *cmd_group;
+
+	if (group == CLOUD_CMD_GROUP_SET) {
+		cmd_group = &set;
+	} else if (group == CLOUD_CMD_GROUP_GET) {
+		cmd_group = &get;
+	} else {
+		return -EINVAL;
+	}
+
+	for (size_t i = 0; i < cmd_group->num_children; i++) {
+		struct cmd rcpt = cmd_group->children[i];
+		recipient_obj = json_object_decode(group_obj,
+						   rcpt.name);
+		if (recipient_obj == NULL) {
+			continue;
+		}
+
+		cmd_parsed.recipient = rcpt.recipient;
+
+		for (size_t j = 0; j < rcpt.num_children; j++) {
+			struct cmd chan = rcpt.children[j];
+			channel_obj = json_object_decode(recipient_obj,
+							 chan.name);
+			if (channel_obj == NULL) {
+				continue;
+			}
+
+			cmd_parsed.channel = chan.channel;
+
+			for (size_t k = 0; k < chan.num_children; k++) {
+				struct cmd typ = chan.children[k];
+				type_obj = json_object_decode(channel_obj,
+							      typ.name);
+
+				if (type_obj == NULL) {
+					continue;
+				}
+
+				cmd_parsed.type = typ.type;
+
+				cloud_command_cb(&cmd_parsed);
+			}
+		}
+	}
```

Comment:
Suspect these nested `for` loops could be removed by using some sort of recursive DFS magic, but not a big deal

Reply:
I agree. I'll leave it for now if this is seen as a non-issue.